geojson-tidy.js

```diff
@@ -97,26 +96,42 @@ function tidy(geojson, options) {
                 }
 
             }
+            keepIdxs.push(i)
+        }
 
-            // Copy the point and timestamp to the tidyOutput
-            tidyOutput.features[tidyOutput.features.length - 1].geometry.coordinates.push(lineString[i]);
-            if (timeStamp) {
-                tidyOutput.features[tidyOutput.features.length - 1].properties.coordTimes.push(timeStamp[i]);
-            }
 
-            // If feature exceeds maximum points, start a new feature beginning at the previuos end point
-            if (tidyOutput.features[tidyOutput.features.length - 1].geometry.coordinates.length % filter.maximumPoints === 0) {
-                tidyOutput.features.push(clone(emptyFeature));
-                tidyOutput.features[tidyOutput.features.length - 1].geometry.coordinates.push(lineString[i]);
-                if (timeStamp) {
-                    tidyOutput.features[tidyOutput.features.length - 1].properties.coordTimes.push(timeStamp[i]);
-                }
+        // If we have > maximumPoints points, take a random sample.
+        // Otherwise just return the entire set
+        if (keepIdxs.length > filter.maximumPoints) {
+            // Randomly remove points until we hit maxLength
+
+            // Split off the first and last indices as we always want to keep these
+            const firstIdx = keepIdxs[0]
+            const lastIdx = keepIdxs[keepIdxs.length - 1]
+            keepIdxs = keepIdxs.slice(1, -2)
+            
+            // until we hit the desired length:
+            while (keepIdxs.length > (filter.maximumPoints - 2)) {
+                // randomly select one of the elements to remove
+                removeIdx = Math.floor(Math.random() * keepIdxs.length)
+                keepIdxs = keepIdxs.filter((element, index) => index !== removeIdx)
             }
```

Comment:
because of `.filter` this is O(n) to remove one point.

I would suggest either use a set for O(1) removals or generate all the points and filter them out in one go

Reply:
why not do something equivalent to
```
const pointsToRemoveSet = new Set(...) // populate somehow

// at the very end call .filter once
keepIdxs = keepIdxs.filter(element => !pointsToRemoveSet.has(element));
```

I think this would be much cheaper in the end.

I am just thninking what if there are 1000 points and we need to remove all but 100, we cannot afford to make that many .filter calls

Reply:
`also have to keep track of which ones you have already "removed".`

not necessarily, I think this can be done efficiently too. For example, shuffle array and take first X points as points to keep/to remove